File: 152-maximum-product-subarray/maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int res=INT_MIN;
        int n=nums.size();
        int left=1;
        int right=1;
        for(int i=0;i<n;i++){
            left*=nums[i];
            res=max(res,left);
            if(left==0){
                left=1;
            }

            right*=nums[n-i-1];
            res=max(res,right);
            if(right==0){
                right=1;
            }
        }
        return res;
    }
};

File: 152-maximum-product-subarray/maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int prod=  Integer.MIN_VALUE;
        int pref = 1;
        int suf = 1;
        int n = nums.length;
        for(int i = 0;i<n;i++){
            pref = pref * nums[i];
            prod = Math.max(pref,prod);
            if(pref == 0){
                pref = 1;
            }
            suf = suf * nums[n-i-1];
            prod = Math.max(suf,prod);
            if(suf == 0){
                suf = 1;
            }
        }
        return prod;
    }
}